library/lua/dfhack.lua

```diff
@@ -162,21 +162,28 @@ NEWLINE = "\n"
 COMMA = ","
 PERIOD = "."
 
+-- calls elem_cb(k, v) for each element of the table
+-- returns true if we iterated successfully, false if not
+local function safe_iterate(table, iterate_fn, elem_cb)
+    local function iterate()
+        for k,v in iterate_fn(table) do elem_cb(k, v) end
+    end
+    return pcall(iterate)
+end
+
+local function print_element(k, v)
+    dfhack.println(string.format("%-23s\t = %s", tostring(k), tostring(v)))
+end
+
 function printall(table)
-    local ok,f,t,k = pcall(pairs,table)
-    if ok then
-        for k,v in f,t,k do
-            print(string.format("%-23s\t = %s",tostring(k),tostring(v)))
-        end
+    if not safe_iterate(table, pairs, print_element) then
+        dfhack.println(tostring(table))
```

Comment:
Just wanted to clarify that `printall()` previously didn't print anything if `pairs()` failed, so we should keep that behavior, and this appears to have changed it (I think). `~` in the Lua interpreter is a combination of `print()` and `printall()`, so I'm sometimes getting duplicate output with this change:

```
[DFHack]# lua ~_VERSION
Lua 5.3
Lua 5.3
```

Reply:
yeah, I was attempting to "fix" its failure to output anything for non-iterables. I thought it was a deficiency that `printall()` didn't output anything for strings. I [noticed](https://github.com/DFHack/dfhack/pull/1917#discussion_r684580485) that it double-printed for strings when using `~` on the interactive prompt, but I wasn't sure how best to handle that. I'll change it back to the original behavior, though.